src/query/ts/m3db/encoded_step_iterator_generic.go

```diff
@@ -88,43 +101,103 @@ func (it *encodedStepIterWithCollector) nextForStep(
 		// If this datapoint is before the current timestamp, add it as a
 		// consolidation candidate.
 		if !dp.Timestamp.After(stepTime) {
-			it.seriesPeek[i].started = false
-			it.collector.AddPointForIterator(dp, i)
+			peek.started = false
+			collector.AddPoint(dp)
 		} else {
 			// This point exists further than the current step.
 			// Set peeked value to this point, then consolidate the retrieved
 			// series.
-			it.seriesPeek[i].point = dp
-			it.seriesPeek[i].started = true
-			return
+			peek.point = dp
+			peek.started = true
+			return peek, collector, nil
 		}
 	}
 
-	if err := iter.Err(); err != nil {
-		it.err = err
+	peek.finished = true
+	return peek, collector, iter.Err()
+}
+
+func (it *encodedStepIterWithCollector) nextParallel() error {
+	var (
+		stepTime = it.stepTime
+
+		multiErr xerrors.MultiError
+	)
+
+	it.wg.Add(len(it.seriesIters))
+	for i := range it.seriesIters {
+		var (
+			i         = i
+			peek      = it.seriesPeek[i]
+			iter      = it.seriesIters[i]
+			collector = it.seriesCollectors[i]
+		)
+
+		it.workerPool.Go(func() {
+			peek, collector, err := nextForStep(peek, iter, collector, stepTime)
```

Comment:
Nice this is a solid approach.

I wonder if it would also be a good idea to basically do a bunch in parallel (like the next X points defined by a constant, like 10 or 20) and then read from intermediary storage after?  That way we don't have to use tons of memory decoding everything up front but every time we do a decode step, we can hot loop the decoding for a little while.  Would also help with less context switching (using goroutines to do decoding, then waiting on result less often).

Basically exactly what you're doing here in batch.